components/proxy/src/main/java/com/hotels/styx/routing/handlers/StandardHttpPipeline.java

```diff
@@ -78,6 +80,9 @@ public StandardHttpPipeline(List<HttpInterceptor> interceptors, HttpHandler hand
 
         @Override
         public StyxObservable<HttpResponse> proceed(HttpRequest request) {
+
+            CurrentRequestTracker.INSTANCE.trackRequest(request);
+
             if (index < interceptors.size()) {
                 HttpInterceptor.Chain chain = new HttpInterceptorChain(this, index + 1);
                 HttpInterceptor interceptor = interceptors.get(index);
```

Comment:
Hi Salah,

Apologies for getting back to you this late. 

I think there is a potential for a race condition. Let me explain.

* HttpPipelineHandler processes its events in a Netty IO thread. But not all plugins necessarily run on the Netty IO thread. Some of them may run asynchronously in a separate worker thread.

* `HttpPipelineHandler` enters in `WAITING_FOR_RESPONSE` right after it has submitted a received request to Styx HTTP pipeline. However a possible asynchronous plugin might still be processing the request on a different thread. And we'd like to show the stack trace of that plugin.

* A stack trace should be displayed *until* the request has traversed thought the entire styx pipeline. But `WAITING_FOR_RESPONSE` state is not necessarily a guarantee that it has done so.

* The stack trace should be withdrawn only after the request has left the HTTP pipeline.

To address this issue, you need to call the `CurrentRequestTracker` from the tail end of the pipeline with a special method to indicate that the request has been fully processed. For example, in `StandardHttpPipeline.proceed`:

```
        @Override
        public StyxObservable<HttpResponse> proceed(HttpRequest request) {

            CurrentRequestTracker.INSTANCE.trackRequest(request);

            if (index < interceptors.size()) {
                HttpInterceptor.Chain chain = new HttpInterceptorChain(this, index + 1);
                HttpInterceptor interceptor = interceptors.get(index);

                try {
                    return interceptor.intercept(request, chain);
                } catch (Throwable e) {
                    return StyxObservable.error(e);
                }
            }

            // HERE LIKE SO:
            // Notify the tracker that a request has traversed through the pipeline
            CurrentRequestTracker.INSTANCE.trackRequestSent(request);   

            return fromRxObservable(toRxObservable(client.handle(request, this.context))
                    .compose(StandardHttpPipeline::sendErrorOnDoubleSubscription));
        }
```

So I would add a special method to mark the request's departure from the pipeline. This would guarantee that the stack trace is displayed as long as necessary. But not too long.